src/app/parcels/add/AddParcelForm.tsx

```diff
@@ -0,0 +1,194 @@
+import { insertParcel } from "@/app/parcels/add/databaseFunctions";
+import { Button } from "@mui/material";
+
+interface AddParcelFields {
+    voucherNumber: string;
+    packingDate: Date | null;
+    timeOfDay: Date | null;
+    shippingMethod: string | null;
+    collectionDate: Date | null;
+    collectionTime: Date | null;
+    collectionCentre: string | null;
+}
+
+const withCollectionFormSections = [
+    VoucherNumberCard,
+    PackingDateCard,
+    TimeOfDayCard,
+    ShippingMethodCard,
+    CollectionDateCard,
+    CollectionTimeCard,
+    CollectionCentreCard,
+];
+
+const noCollectionFormSections = [
+    VoucherNumberCard,
+    PackingDateCard,
+    TimeOfDayCard,
+    ShippingMethodCard,
+];
+
+const initialFields: AddParcelFields = {
+    voucherNumber: "",
+    packingDate: null,
+    timeOfDay: null,
+    shippingMethod: "",
+    collectionDate: null,
+    collectionTime: null,
+    collectionCentre: null,
+};
+
+const initialFormErrors: FormErrors = {
+    voucherNumber: Errors.none,
+    packingDate: Errors.initial,
+    timeOfDay: Errors.initial,
+    shippingMethod: Errors.initial,
+    collectionDate: Errors.initial,
+    collectionTime: Errors.initial,
+    collectionCentre: Errors.initial,
+};
+
+const AddParcelForm: React.FC<{ id: string }> = (props: { id: string }) => {
+    const router = useRouter();
+    const [fields, setFields] = useState(initialFields);
+    const [formErrors, setFormErrors] = useState(initialFormErrors);
+    const [submitError, setSubmitError] = useState(Errors.none);
+    const [submitErrorMessage, setSubmitErrorMessage] = useState("");
+    const [submitDisabled, setSubmitDisabled] = useState(false);
+
+    const formSections =
+        fields.shippingMethod === "Collection"
+            ? withCollectionFormSections
+            : noCollectionFormSections;
+
+    const fieldSetter = setField(setFields, fields);
+    const errorSetter = setError(setFormErrors, formErrors);
+
+    const mergeDateAndTime = (date: Date, time: Date): Date => {
+        return new Date(
+            date.getFullYear(),
+            date.getMonth(),
+            date.getDate(),
+            time.getHours(),
+            time.getMinutes()
+        );
+    };
+
+    const submitForm = async (): Promise<void> => {
+        setSubmitErrorMessage("");
+        setSubmitDisabled(true);
+        let inputError;
+        if (fields.shippingMethod === "Collection") {
+            inputError = checkErrorOnSubmit(formErrors, setFormErrors);
+        } else {
+            inputError = checkErrorOnSubmit(formErrors, setFormErrors, [
+                "voucherNumber",
+                "packingDate",
+                "timeOfDay",
+                "shippingMethod",
+            ]);
+        }
+        if (inputError) {
+            setSubmitError(Errors.submit);
+            setSubmitDisabled(false);
+            return;
+        }
+        if (fields.packingDate === null || fields.timeOfDay === null) {
```

Comment:
So this should never happen. I will set it to set an error message if it ever does happen, but `checkErrorOnSubmit` should detect if these values do end up being null.

Reply:
(Without it the types break when merging the date and times)

Reply:
if you can guarantee that `packingDate` and `timeOfDay` will be non-null at this point (because of `checkErrorOnSubmit`) you can make the type system aware with an exclamation mark - `fields.packingDate!` will convert it from type `Date | null` to `Date` (with the idea that if it was null an error is thrown)

Reply:
Good point! Will change now.